src/Proxies/LegacyProxy.php

```diff
@@ -0,0 +1,92 @@
+<?php
+/**
+ * LegacyProxy class file.
+ *
+ * @package Automattic/WooCommerce/Tools/Proxies
+ */
+
+namespace Automattic\WooCommerce\Proxies;
+
+use \Psr\Container\ContainerInterface as Container;
+
+/**
+ * Proxy class to access legacy WooCommerce functionality.
+ *
+ * This class should be used to interact with code outside the `src` directory, especially functions and classes
+ * in the `includes` directory, unless a more specific proxy exists for the functionality at hand (e.g. `ActionsProxy`).
+ * Idempotent functions can be executed directly.
+ *
+ * @package Automattic\WooCommerce\Tools\Proxies
+ */
+class LegacyProxy {
+
+	/**
+	 * Gets an instance of a given legacy class.
+	 * This must not be used to get instances of classes in the `src` directory.
+	 *
+	 * @param string $class_name The name of the class to get an instance for.
+	 * @param mixed  ...$args Parameters to be passed to the class constructor or to the appropriate internal 'get_instance_of_' method.
+	 *
+	 * @return object The instance of the class.
+	 * @throws \Exception The requested class belongs to the `src` directory, or there was an error creating an instance of the class.
+	 */
+	public function get_instance_of( string $class_name, ...$args ) {
+		if ( false !== strpos( $class_name, '\\' ) ) {
+			throw new \Exception( 'The LegacyProxy class is not intended for getting instances of classes in the src directory, please use constructor injection or the instance of \\Psr\\Container\\ContainerInterface for that.' );
+		}
+
+		try {
+			// If a class has a dedicated method to obtain a instance, use it.
+			$method = 'get_instance_of_' . strtolower( $class_name );
+			if ( method_exists( __CLASS__, $method ) ) {
+				return $this->$method( ...$args );
+			}
+
+			// If the class is a singleton, use the "instance" method.
+			if ( method_exists( $class_name, 'instance' ) ) {
+				return $class_name::instance();
+			}
+
+			// Fallback to simply creating a new instance of the class.
+			return new $class_name( ...$args );
+		} catch ( \Exception $e ) {
+			throw new \Exception( __CLASS__ . '::' . __FUNCTION__ . ": error when getting a new instance of $class_name: {$e->getMessage()}" );
```

Comment:
We are swallowing the Exception type here, let's throw the original exception.

Reply:
I wanted to provide some context to help determine. How can we throw the original exception but providing more information?

Reply:
As discussed, the `try` block can be simply removed, also `call_function´ and `call_static` in the same class don't try to catch exceptions either.